### PR TITLE
Update to Diesel master

### PR DESCRIPTION
**Pull request by sgrif**
Diesel's current master branch is likely to be reasonably representative
of what 1.0 will look like from an API perspective. If there's any other
parts of the code that we dislike and want to improve the API for, now
is the time to speak up!

The test changes are due to some deadlocks that I'm seeing now (which
should have always been happening, I suspect some perf characteristic
has changed somewhere)

### CODE REVIEW

File: src/bin/update-downloads.rs [hunk 1]
```diff
@@ -93,23 +80,23 @@ fn collect(conn: &PgConnection, rows: &[VersionDownload]) -> QueryResult<()> {
...
-            (crate_downloads::crate_id, crate_downloads::date),
-            do_update().set(crate_downloads::downloads.eq(crate_downloads::downloads + amt)),
-        )).into(crate_downloads::table)
+        insert_into(crate_downloads::table)
+            .values((
+                crate_downloads::crate_id.eq(crate_id),
+                crate_downloads::downloads.eq(amt),
+                crate_downloads::date.eq(download.date),
+            ))
+            .on_conflict(crate_downloads::table.primary_key())
```
Comment (sgrif) [inline src/bin/update-downloads.rs]: Indentation is so much better now...

File: src/schema.rs [hunk 2]
```diff
@@ -783,3 +783,26 @@ joinable!(version_authors -> users (user_id));
 joinable!(version_authors -> versions (version_id));
 joinable!(version_downloads -> versions (version_id));
 joinable!(versions -> crates (crate_id));
+
+allow_tables_to_appear_in_same_query!(
```
Comment (sgrif) [inline src/schema.rs]: Hopefully this is less confusing now

Comment (carols10cents) [reply src/schema.rs]: If we generate it for all tables, why do we have to explicitly list this at all...?

File: src/krate/mod.rs [hunk 3]
```diff
@@ -280,20 +279,16 @@ impl<'a> NewCrate<'a> {
 }
 
 impl Crate {
```
Comment (vignesh-sankaran) [inline src/krate/mod.rs]: Could you add doc comments to the `by_name`, `with_name`, and `all` functions in this struct?

File: src/schema.rs (hunk 2 above)
Comment (sgrif) [reply src/schema.rs]: I don't understand the question

Comment (carols10cents) [reply src/schema.rs]: > we just speculatively generate it for all tables

By "we", do you mean "crates.io" or "diesel"? If Diesel is doing this speculative generation, why does it have to be listed explicitly in src/schema.rs at all? Why isn't this built-in to diesel?

Comment (sgrif) [reply src/schema.rs]: This file was generated by `diesel print-schema`

**Comment by sgrif** [comment]
I've rebased and killed the remaining raw inserts in `update-downloads.rs`. I'm going to leave the ones in `category.rs` for now, as they're a bit verbose and this PR is already really big

**Comment by carols10cents** [comment]
Working ok for me! Let's do this!

bors: r+

**Comment by carols10cents** [comment]
oops

**Comment by carols10cents** [comment]
bors: r+

**Comment by carols10cents** [comment]
i'm getting real tired of that intermittent faker error...

bors: r+

**Comment by sgrif** [comment]
bors: retry

Looks like it's some NPM issue

**Comment by sgrif** [comment]
I guess retry doesn't work?

bors: r+

**Comment by carols10cents** [comment]
bors: r+

**Comment by sgrif** [comment]
JavaScript is the worst.

**Pull request merged**